decisions/2021/04/13-fe-initiatives.md

```diff
@@ -0,0 +1,103 @@
+When: April 13, 2021 @ 15:00 CET
+Where: Online
+Attending: Austin, Viktor, Bjørn, Christos, Debora, Edoardo, Ismay,
+Jan-Gerke, Joakim, Joe, Kai, Martin, Simona, Hendrik, Médi-Rémi,
+Jennifer, Jan Henrik, Birk
+
+# Agenda
+
+## Introductions!
+
+  - Welcome Simona!
+
+## Decisions
+
+### Standardize application file layout #248 
+    
+Discussion: https://github.com/dhis2/notes/discussions/257
+
+Goal was unambiguous rules, difficult to define across all
+application sizes.
+
+Guidelines: 
+
+1. Have similar organisation strategies between all our apps
+2. Organise by-function, not by-type
+3. Avoid ambiguity (i.e. no conceptual overlap in folder names)
+4. Avoid generic folder names, and attempt to chose names that
+   immediately reveal what’s inside
+5. Accept that a project-structure is not a static thing you can
+   define up-front, the structure should evolve with the app
+
+Rules:
+
+1. Routes live under `src/pages`
+2. Components live under `src/components` and must only contain JSX
```

Comment:
I'm wondering if having a folder that _can't_ contain jsx and only regular js would be useful as well. For the kind of code that often lives in utils, services, etc.

Reply:
I was actually thinking - JSX can ONLY be used in `src/components` (and `src/pages`).  Any other directories should therefore be JSX-free.  I don't think we can say "only" JSX in `src/component`, as there will definitely be some javascript in there too, but we can enforce that no JSX is allowed outside those two directories, and can employ the convention that all js files in `src/components` should export a single React component?

Reply:
Yeah, this can definitely be phrased a lot clearer.

Reply:
@amcgee, yeah I think there are several eslint/react rules we could use there. Restricting jsx to the components dir, single component per file, etc. Sounds useful 👍

Reply:
This is interesting, because I had a different understanding of what we were talking about yesterday.
In my opinion, depending on the scale, a single `components` folder on the root level might not make much sense and could lead to unnecessary abstractions (which is what @ismay worried about in the beginning of the discussion, and I agree; see sms config app - initially had a `src/components` folder but moving away from that removed a lot of complexity)

Similarly having a fixed `src/pages` folder to me only makes sense IF we use `next.js`. Otherwise we'll introduce overhead to the app. If we approach this the other way: "No next.js-like pages folder until using next.js", it's quite trivial work to restructure an app when introducing next.js (or a router service) while we reduce the complexity of all apps that don't use next.js/our own implementaiton/routing-service (which is all apps right now and - until finalized - still a vague idea => If we use a next.js architecture now, but end up with a slightly different implementation, we have more work because we first have to restructure apps to next.js-style and then additionally adjust to our dhis2-adjusted variant. If we do this once we have the dhis2-variant, it'd be the same amount of work as we'd be doing now, so there'd be no time saved by doing this now (enforcing a rule) but on the other hand, time save if we defer this until we know exactly how we want to do this).

Also having a `components` folder is actually the opposite of organizing by function, it's organizing by type (component vs non-component). So I'm against having a rule for this. There are def. apps where this makes sense, e. g. with a single-domain-complexity, like the scheduler app, but one things get more complicated, like the maintenance app, a `src/components` folder would introduce unnecessary complexity (which is why we struggled having rules in the first place).

What I understood yesterday is: Different modules should separate code into components and non-components (which makes a lot of since for internal module organization), but we should keep this as a guide, rather than a rule, because different apps need different levels of complexity

Reply:
Made a suggestion to reflect this approach

Reply:
:+1: for suggestions.